src/snowflake/cli/plugins/rpc/manager.py

```diff
@@ -0,0 +1,62 @@
+# Copyright (c) 2024 Snowflake Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from __future__ import annotations
+
+import importlib
+import inspect
+import pkgutil
+from dataclasses import dataclass
+from typing import Optional
+
+from pygls.server import LanguageServer
+
+
+@dataclass
+class ConnectionParams:
+    session_token: Optional[str]
+    master_token: Optional[str]
+    account: Optional[str]
+
+
+class RpcManager:
+    """
+    Base class LSP Controller
+    """
+
+    def __init__(self):
+        super().__init__()
+        server = LanguageServer(name="lsp_controller", version="v0.0.1")
+
+        plugins = self.discover_lsp_plugins("snowflake.cli.plugins")
+        for plugin_func in plugins:
+            plugin_func(server)
+        server.start_io()
+
+    def discover_lsp_plugins(self, package_name):
+        plugins = []
+        package = importlib.import_module(package_name)
+        for _, module_name, is_pkg in pkgutil.iter_modules(package.__path__):
+            if is_pkg:
+                submodule = importlib.import_module(f"{package_name}.{module_name}")
+                for _, submodule_name, is_pkg in pkgutil.iter_modules(
+                    submodule.__path__
+                ):
+                    module = importlib.import_module(
+                        f"{package_name}.{module_name}.{submodule_name}"
+                    )
+                    for _name, obj in inspect.getmembers(module, inspect.isfunction):
+                        if hasattr(obj, "_lsp_context"):
+                            plugins.append(obj)
+        return plugins
```

Comment:
@sfc-gh-turbaszek I've tried to use Pluggy to replace this discovery mechanism and I am stumped. I tried following a small toy example of it from their [docs](https://pluggy.readthedocs.io/en/stable/#a-toy-example) but I don't understand how the automatic plugin discovery is supposed to work, because it seems like it either needs some mapping the way the command_plugins_loader.py has get_builtin_plugin_name, or it uses the setuptools_entrypoints which I don't think I should need since there's no need for external plugins here.

Reply:
It looks like SnowCLI uses the hook markers but then uses its own mechanism to identify the plugins. The LoadedCommandPlugin has a command_spec that gets used in "_load_command_spec" in command_plugins_loader to return the plugin.command_spec(), and the contents of that occur in plugin_spec.py for each plugin. The command_spec() for native apps for example is:
```
@plugin_hook_impl
def command_spec():
    return CommandSpec(
        parent_command_path=SNOWCLI_ROOT_COMMAND_PATH,
        command_type=CommandType.COMMAND_GROUP,
        typer_instance=commands.app.create_instance(),
    )
```
And it seems like the @plugin_hook_impl from the pluggy marker is not really used/needed by the builtin plugin process since the code doesn't use pluggy for discovery of those builtins. e.g., removing @plugin_hook_impl, everything still works. 

So I guess I can follow the pattern there for command_spec but with another spec for the LSP stuff.

Reply:
Using pluggy seemed more official and in line with the code in the repo than the way I was doing it before. Did you prefer the previous version that didn't use pluggy? In the end SnowCLI has its own mechanism to find the plugins; the plugin.command_spec and plugin.lsp_spec, so it isn't automatic discovery for the internal plugins. That's my current understanding.